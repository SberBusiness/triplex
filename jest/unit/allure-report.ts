import {JasmineAllureReporter} from 'allure-jasmine';
import {Status, TestResult} from 'allure-js-commons';

const reporter = new JasmineAllureReporter({
    resultsDir: 'tests/out/unit/allure/allure-results',
    testMapper: (result: TestResult) => {
        if (result.status == Status.SKIPPED) result.fullName = `(WAS SKIPPED) ${result.fullName}`;
        return result;
    },
});

// @ts-ignore
jasmine.getEnv().addReporter(reporter);

export const allure = reporter.getInterface();
