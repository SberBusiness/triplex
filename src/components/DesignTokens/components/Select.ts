import {TDesignTokenValue, TDesignTokenValues} from '@sberbusiness/triplex/components/DesignTokens/types/DesignTokenTypes';

// Название токенов компонента Select.
export const designTokensComponentsSelectKeys = [
    'Background_Default',
    'Background_Disabled',

    'Caret_Color_Default',
    'Caret_Color_Disabled',

    'Border_Color_Default',
    'Border_Color_Disabled',
    'Border_Color_Error',
    'Border_Color_Focus',

    'Color_Default',
    'Color_Disabled',
    'Color_Placeholder',
] as const;
// Тип, содержащий названия токенов компонента Select.
export type TDesignTokensComponentsSelectKeys = (typeof designTokensComponentsSelectKeys)[number];
// Тип, содержащий названия токенов компонента Select и их значения.
export type TDesignTokensComponentsSelectValue = Record<TDesignTokensComponentsSelectKeys, TDesignTokenValue>;
// Тип, содержащий названия токенов компонента Select и их значения в светлой и темной теме.
export type TDesignTokensComponentsSelectValues = Record<TDesignTokensComponentsSelectKeys, TDesignTokenValues>;
// Тип локальных токенов компонента Select.
export type TDesignTokensComponentsSelect = {Select: TDesignTokensComponentsSelectValue};

// Токены компонента Select в светлой и темной темах.
export const Select_Tokens: TDesignTokensComponentsSelectValues = {
    Background_Default: [{ref: 'Basic.100'}, {value: 'none'}], // var(--triplex-Select-Background_Default)
    Background_Disabled: [{ref: 'Neutral.100'}, {ref: 'Basic.700'}], // var(--triplex-Select-Background_Disabled)

    Border_Color_Default: [{ref: 'Neutral.500'}, {ref: 'Neutral.900'}], // var(--triplex-Select-Border_Color_Default)
    Border_Color_Disabled: [{ref: 'Neutral.500'}, {ref: 'Basic.300'}], // var(--triplex-Select-Border_Color_Disabled)
    Border_Color_Error: [{ref: 'Error.700'}, {ref: 'Error.500'}], // var(--triplex-Select-Border_Color_Error)
    Border_Color_Focus: [{ref: 'Basic.700'}, {ref: 'Neutral.700'}], // var(--triplex-Select-Border_Color_Focus)

    Caret_Color_Default: [{ref: 'Basic.700'}, {ref: 'Neutral.700'}], // var(--triplex-Select-Caret_Color_Default)
    Caret_Color_Disabled: [{ref: 'Neutral.500'}, {ref: 'Basic.300'}], // var(--triplex-Select-Caret_Color_Disabled)

    Color_Default: [{ref: 'Basic.700'}, {ref: 'Neutral.100'}], // var(--triplex-Select-Color_Default)
    Color_Disabled: [{ref: 'Neutral.900'}, {ref: 'Basic.300'}], // var(--triplex-Select-Color_Disabled)
    Color_Placeholder: [{ref: 'Neutral.900'}, {ref: 'Neutral.900'}], // var(--triplex-Select-Color_Placeholder)
};
