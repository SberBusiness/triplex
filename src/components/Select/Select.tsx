import React from 'react';
import {EInputGroupPosition} from '@sberbusiness/triplex/components/InputGroup/InputGroup';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {ISelectBaseOption, ISelectBaseProps, SelectBase} from '@sberbusiness/triplex/components/SelectBase/SelectBase';

/** Свойства опции списка. */
export interface ISelectOption extends ISelectBaseOption {}

/** Свойства компонента Select.*/
export interface ISelectProps extends ISelectBaseProps {
    /** Позиция внутри компонента InputGroup. */
    groupPosition?: EInputGroupPosition;
}

const mapInputGroupPositionToCSSClass = {
    [EInputGroupPosition.LEFT]: 'cssClass[left]',
    [EInputGroupPosition.INTERMEDIATE]: 'cssClass[intermediate]',
    [EInputGroupPosition.RIGHT]: 'cssClass[right]',
};

/**
 * Компонент Select.
 * В качестве value и options принимает объекты типа ISelectOption.
 * Если требуется кастомизация options или другой формат value – создайте свой на основе SelectExtended.
 */
export const Select: React.FC<ISelectProps> = ({className, groupPosition, targetProps, ...selectBaseProps}) => {
    const targetClassName = classnames(
        'cssClass[selectTarget]',
        targetProps?.className,
        groupPosition && mapInputGroupPositionToCSSClass[groupPosition]
    );
    const selectClassName = classnames('cssClass[select]', {'cssClass[grouped]': !!groupPosition}, className);

    return (
        <SelectBase
            className={selectClassName}
            targetProps={{
                ...targetProps,
                className: targetClassName,
            }}
            dropdownListItemClassName="cssClass[selectDropdownListItem]"
            {...selectBaseProps}
        />
    );
};
