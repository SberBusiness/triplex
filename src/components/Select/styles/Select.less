@import 'src/styles/style';

.select {
    min-width: 72px;
}

.selectDropdownListItem {
    overflow: hidden;
    text-overflow: ellipsis;
}

.grouped {
    flex: 1;

    &:not(:first-child) {
        margin-left: -1px;
    }

    .left {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
    }

    .intermediate {
        border-radius: 0;
    }

    .right {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
    }

    .selectTarget.error {
        z-index: 1;
    }

    .selectTarget:focus {
        z-index: 2;
    }
}
