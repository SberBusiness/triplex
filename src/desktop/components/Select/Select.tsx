import React from 'react';
import {EInputGroupPosition} from '@sberbusiness/triplex/desktop/components/InputGroup/InputGroup';
import {classnames} from '@sberbusiness/triplex/common/utils/classnames/classnames';
import {ISelectBaseOption, ISelectBaseProps, SelectBase} from '@sberbusiness/triplex/desktop/components/SelectBase/SelectBase';

/** Свойства опции списка. */
export interface ISelectOption extends ISelectBaseOption {}

/** Свойства компонента Select.*/
export interface ISelectProps extends ISelectBaseProps {
    /** Позиция внутри компонента InputGroup. */
    groupPosition?: EInputGroupPosition;
}

const mapInputGroupPositionToCSSClass = {
    [EInputGroupPosition.LEFT]: 'cssClass[left]',
    [EInputGroupPosition.INTERMEDIATE]: 'cssClass[intermediate]',
    [EInputGroupPosition.RIGHT]: 'cssClass[right]',
};

/**
 * Компонент Select.
 * В качестве value и options принимает объекты типа ISelectOption.
 * Если требуется кастомизация options или другой формат value - создайте CustomSelect на основе SelectExtended.
 */
export const Select: React.FC<ISelectProps> = ({className, groupPosition, ...selectBaseProps}) => {
    const targetClassName = classnames('cssClass[selectTarget]', groupPosition && mapInputGroupPositionToCSSClass[groupPosition]);
    const selectClassName = classnames('cssClass[select]', {'cssClass[grouped]': !!groupPosition}, className);

    return (
        <SelectBase
            className={selectClassName}
            targetClassName={targetClassName}
            dropdownListItemClassName="cssClass[selectDropdownListItem]"
            {...selectBaseProps}
        />
    );
};
