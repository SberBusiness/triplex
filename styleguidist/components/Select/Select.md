```jsx
const [value, setValue] = React.useState();

const options = [
    {value: 'i1', label: 'Первый'},
    {value: 'i2', label: 'Второй'},
    {value: 'i3', label: 'Третий'},
    {value: 'i4', label: 'Четвертый'},
    {value: 'i5', label: 'Пятый'},
    {value: 'i6', label: 'Шестой'},
    {value: 'i7', label: 'Седьмой'},
    {value: 'i8', label: 'Восьмой'},
    {value: 'i9', label: 'Девятый'},
    {value: 'i10', label: 'Десятый'},
];

<Select
    value={value}
    options={options}
    placeholder="Выберите значение"
    mobileTitle="Выберите значение"
    onChange={setValue}
/>
```

### Disabled state

```jsx
const [value, setValue] = React.useState();

const options = [
    {value: 'i1', label: 'Первый'},
    {value: 'i2', label: 'Второй'},
    {value: 'i3', label: 'Третий'},
    {value: 'i4', label: 'Четвертый'},
    {value: 'i5', label: 'Пятый'},
    {value: 'i6', label: 'Шестой'},
    {value: 'i7', label: 'Седьмой'},
    {value: 'i8', label: 'Восьмой'},
    {value: 'i9', label: 'Девятый'},
    {value: 'i10', label: 'Десятый'},
];

<Select
    value={value}
    options={options}
    placeholder="Выберите значение"
    mobileTitle="Выберите значение"
    onChange={setValue}
    disabled
/>
```

### Error state

```jsx
import {AlertContext} from '@sberbusiness/triplex/components/Alert/AlertContext/AlertContext';
import {EAlertType} from '@sberbusiness/triplex/components/Alert/EAlertType';
import {Gap} from '@sberbusiness/triplex/components/Gap/Gap';

const [value, setValue] = React.useState();

const options = [
    {value: 'i1', label: 'Первый'},
    {value: 'i2', label: 'Второй'},
    {value: 'i3', label: 'Третий'},
    {value: 'i4', label: 'Четвертый'},
    {value: 'i5', label: 'Пятый'},
    {value: 'i6', label: 'Шестой'},
    {value: 'i7', label: 'Седьмой'},
    {value: 'i8', label: 'Восьмой'},
    {value: 'i9', label: 'Девятый'},
    {value: 'i10', label: 'Десятый'},
];

<>
    <Select
        value={value}
        options={options}
        placeholder="Выберите значение"
        mobileTitle="Выберите значение"
        onChange={setValue}
        error
        targetProps={{
            'aria-invalid': true,
            'aria-errormessage': 'select-error-id',
        }}
    />
    <Gap size={8} />
    <AlertContext id="select-error-id" type={EAlertType.ERROR}>Описание ошибки</AlertContext>
</>
```

### Loading state

```jsx
const [value, setValue] = React.useState();

const options = [
    {value: 'i1', label: 'Первый'},
    {value: 'i2', label: 'Второй'},
    {value: 'i3', label: 'Третий'},
    {value: 'i4', label: 'Четвертый'},
    {value: 'i5', label: 'Пятый'},
    {value: 'i6', label: 'Шестой'},
    {value: 'i7', label: 'Седьмой'},
    {value: 'i8', label: 'Восьмой'},
    {value: 'i9', label: 'Девятый'},
    {value: 'i10', label: 'Десятый'},
];

<Select
    value={value}
    options={options}
    placeholder="Выберите значение"
    mobileTitle="Выберите значение"
    onChange={setValue}
    loading
/>
```

### With label

```jsx
import {Col} from '@sberbusiness/triplex/components/Col/Col';
import {Label} from '@sberbusiness/triplex/components/Label/Label';
import {Field} from '@sberbusiness/triplex/components/Field/Field';

const [value, setValue] = React.useState();

const options = [
    {value: 'i1', label: 'Яблоко'},
    {value: 'i2', label: 'Апельсин'},
    {value: 'i3', label: 'Лимон'},
    {value: 'i4', label: 'Груша'},
    {value: 'i5', label: 'Манго'},
    {value: 'i6', label: 'Авокадо'},
    {value: 'i7', label: 'Ананас'},
    {value: 'i8', label: 'Банан'},
    {value: 'i9', label: 'Гранат'},
    {value: 'i10', label: 'Киви'},
];

<Field alignLabel>
    <Col size={3}>
        <Label>
            <Label.Text id="select_id1">Любимый фрукт</Label.Text>
        </Label>
    </Col>
    <Col size={9}>
        <Select
            aria-labelledby="select_id1"
            value={value}
            options={options}
            placeholder="Выберите значение"
            mobileTitle="Выберите значение"
            onChange={setValue}
        />
    </Col>
</Field>
```
