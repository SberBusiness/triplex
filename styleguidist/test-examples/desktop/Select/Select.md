```jsx 
import React, {useState, useEffect} from 'react';
import ComponentPreview from '../../../common/components/ComponentPreview/ComponentPreview';
import ComponentOptions from '../../../common/components/ComponentOptions/ComponentOptions';
import {SelectExample} from './SelectExample';

const selectDataTestId = 'Default_Select';

const [topOrientation, setTopOrientation] = useState(false);
const [disabled, setDisabled] = useState(false);
const [error, setError] = useState(false);
const [loading, setLoading] = useState(false);
const [isOpen, setIsOpen] = useState(false);
const [inputValue, setInputValue] = useState('');

useEffect(() => {
    if (isOpen) {
        const selectTarget = document.querySelector(`[data-test-id="${selectDataTestId}__target"]`);
        if (selectTarget) {
            selectTarget.click();
        }
    }
}, [isOpen]);

const checkboxOptions = [
    {
        id: 'Select-openToTop',
        label: 'Dropdown to top',
        checked: topOrientation,
        onChange: setTopOrientation,
    }, {
        id: 'Select-disabled',
        label: 'Disabled',
        checked: disabled,
        onChange: setDisabled,
    }, {
        id: 'Select-error',
        label: 'Error',
        checked: error,
        onChange: setError,
    }, {
        id: 'Select-loading',
        label: 'Loading',
        checked: loading,
        onChange: setLoading,
    }, {
        id: 'Select-isOpen',
        label: 'Is open',
        hidden: true,
        checked: isOpen,
        onChange: setIsOpen,
    },
];

const inputOptions = [
    {
        id: 'inputValue',
        hidden: true,
        value: inputValue,
        onChange: setInputValue,
    },
];

<>
    <ComponentOptions checkboxOptions={checkboxOptions} inputOptions={inputOptions}/>
    <ComponentPreview
        style={{
            boxSizing: 'border-box',
            height: isOpen ? '240px' : undefined,
            paddingTop: isOpen && topOrientation ? '208px' : undefined,
        }}
    >
        <SelectExample
            error={error}
            disabled={disabled}
            loading={loading}
            placeholder="Выберите значение" 
            topOrientation={topOrientation}
            inputValue={inputValue}
            data-test-id={selectDataTestId}
        />
    </ComponentPreview>
</>
````