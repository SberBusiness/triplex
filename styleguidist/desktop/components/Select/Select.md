```jsx noeditor
import ComponentStylesDependency from '../../../common/components/ComponentStylesDependency/ComponentStylesDependency';

<ComponentStylesDependency
    componentTitle="Select"
    isMobileComponent={false}
/>
```

```jsx 
import {Select} from '@sberbusiness/triplex/desktop/components/Select/Select';
import {ExampleControlPanel} from '../common/ExampleControlPanel/ExampleControlPanel';

const [value, setValue] = React.useState();
const [loading, setLoading] = React.useState(false);
const [error, setError] = React.useState(false);
const [disabled, setDisabled] = React.useState(false);

const options = [
    {value: 'i1', label: 'Первый'},
    {value: 'i2', label: 'Второй'},
    {value: 'i3', label: 'Третий'},
    {value: 'i4', label: 'Четвертый'},
    {value: 'i5', label: 'Пятый'},
    {value: 'i6', label: 'Шестой'},
    {value: 'i7', label: 'Седьмой'},
    {value: 'i8', label: 'Восьмой'},
    {value: 'i9', label: 'Девятый'},
    {value: 'i10', label: 'Десятый'},
];

const placeholder = 'Выберите значение';

const renderControlPanel = () => (
    <ExampleControlPanel>
        <input type="checkbox" checked={loading} onChange={(event) => setLoading(event.target.checked)} data-label="Loading" />
        <input type="checkbox" checked={error} onChange={(event) => setError(event.target.checked)} data-label="Error" />
        <input type="checkbox" checked={disabled} onChange={(event) => setDisabled(event.target.checked)} data-label="Disabled" />
    </ExampleControlPanel>
);

<>
    {renderControlPanel()}
    <Select
        value={value}
        onChange={setValue}
        options={options}
        placeholder={placeholder}
        loading={loading}
        error={error}
        disabled={disabled}
    />
</>
```
