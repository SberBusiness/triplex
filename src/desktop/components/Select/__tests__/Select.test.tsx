import React from 'react';
import {create} from 'react-test-renderer';
import {ISelectOption, Select} from '../Select';
import {allure} from '@jest/unit/allure-report';

jest.mock('@sberbusiness/icons/CaretdownSrvxIcon16', () => ({
    CaretdownSrvxIcon16: 'svg',
}));

jest.mock('@sberbusiness/icons/SpinnersmallAniIcon20', () => ({
    SpinnersmallAniIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/ClosenotificationSrvxIcon16', () => ({
    ClosenotificationSrvxIcon16: 'svg',
}));

jest.mock('@sberbusiness/triplex/common/utils/uniqueId', () => ({
    uniqueId: jest.fn(() => 0),
}));

describe('Select', () => {
    beforeEach(() => {
        allure.feature('Select');
    });

    it('should render correctly', () => {
        const options: Array<ISelectOption> = [
            {value: 'i1', label: 'Первый'},
            {value: 'i2', label: 'Второй'},
            {value: 'i3', label: 'Третий'},
        ];

        const tree = create(
            <Select data-test-id="Select" options={options} placeholder="Placeholder" value={options[0]} onChange={jest.fn()} />
        );

        expect(tree).toMatchSnapshot();
    });
});
