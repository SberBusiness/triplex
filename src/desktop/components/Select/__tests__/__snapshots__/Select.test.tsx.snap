// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render correctly 1`] = `
<div
  className="cssClass[selectExtended] cssClass[select]"
  data-test-id="Select"
>
  <div
    aria-controls={0}
    aria-expanded={false}
    aria-haspopup="listbox"
    className="cssClass[selectExtendedTarget] cssClass[selectTarget]"
    data-test-id="Select__target"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="combobox"
    tabIndex={0}
  >
    <div
      className="cssClass[label]"
      data-test-id="Select__target__selected_text"
    >
      Первый
    </div>
    <svg
      className="cssClass[caretIcon] hoverable"
    />
  </div>
</div>
`;
